### DISCUSSION TITLE
Feedback on Gradle Versions Catalog support?

### DISCUSSION THREAD

**Discussion by jmfayard** [Splitties/refreshVersions#592] [General]
We have released in 0.50.0 support for Gradle Versions Catalog support. 

--> See https://jmfayard.github.io/refreshVersions/CHANGELOG/

This is a new and huge feature so we need you to test it out and provide feedback.

1. Update Gradle with `./gradlew wrapper --gradle-version 7.5.1` (Versions Catalogs are stable since Gradle 7.4)
2. [Setup refreshVersions 0.50.+ in your build](https://jmfayard.github.io/refreshVersions/setup/) in its latest release (our support for Versions Catalog is very new).
3. Run `./gradlew refreshVersionsMigrate --mode=VersionCatalogAndVersionProperties`
4. See the generate catalog at `gradle/libs.versions.toml`
5. See your `git diff` to check how your build has been migrated

[Read the friendly documentation here](https://github.com/jmfayard/refreshVersions/wiki/RefreshVersions-%E2%99%A5%EF%B8%8F-Gradle-Version-Catalog)

Thanks for your help!

**Comment by jmfayard** [comment]
Issue found here https://github.com/jmfayard/refreshVersions/discussions/396#discussioncomment-3613993

>  Extension of type 'VersionCatalogsExtension' does not exist. Currently registered extension types: [ExtraPropertiesExtension]
> Gradle 7.5.1 and refreshVersions 0.50.0

**EDIT** Fixed in 0.50.1
https://jmfayard.github.io/refreshVersions/CHANGELOG/

**Comment by mcpiroman** [comment]
Hi, I migrated from buildSrc directly to VersionsCatalogOnly and then `refreshVersions` also generates a new `versions.properties`, which I don't use (because there is no point to use both?).

**Reply by jmfayard** [reply]
Hello I guess this will be a frequent question. There are multiple partuak answers.

Here is a short summary:

- I agree in principle but Rom was not built in one day
- you can make the file less visible by diffusing refreshVersions to put the file under `gradle/versions.properties`
- personally I think that refreshVersions dependency notations are a better (complementary) solution for common libraries

See if this answer your questions here

https://github.com/jmfayard/refreshVersions/wiki/RefreshVersions-%E2%99%A5%EF%B8%8F-Gradle-Version-Catalog#faq-are-you-giving-up-on-the-versionsproperties-file-

**Reply by myjohndoeside** [reply]
So, currently this plugin could not work without `versions.properties` file?

**Reply by jmfayard** [reply]
Currently yes. 
Cause that's how refreshVersions work from the start and from before versions catalog.

**Comment by imashnake0** [comment]
Just merged a branch that I was using to try out the feature. [It works flawlessly!](https://github.com/imashnake0/Animite/commit/6192701c807c317aae606b8d8fb64a9a945956ea)

P.S. Sorry I haven't been able to contribute lately, had a lot on my plate :sob:

**Comment by jmfayard** [comment]
I'm glad to hear!

**Comment by Arc-E-Tect** [comment]
I like the fact that Version Catalogs are now supported. Took me a while before I had the time to switch and check this new feature.

I noticed that none of the --mode flags process the plugins section in the TOML file. libraries are processed and an _ is put in place of the version, but with plugins this is not the case. I need to manually replace the version strings with an _ after that all still works.

**Comment by Commodore68** [comment]
Gotta give you kudos for this! being able to use this to migrate to version catalogs was great. Even for a multi-module gradle project that makes heavy use of build-src. The semi-automatic migration was very helpful for dependencies in my main modules.

Then it was just a matter of figuring out how to migrate the things in my build-src over to the version catalog, which I was able to do with a minimal amount of code.

only minor issue I had, was that I had some dependencies using a gradle extension variable in kts build scripts to set versions and the auto migrate did not replace those with _

```kotlin
val versionVariable: String by rootProject.extra //OR val versionVariable: String by project

implementation("group:name:$versionVariable")
```

great work again.

**Comment by dmcg** [comment]
In my experimental project `./gradlew refreshVersionsMigrate  --mode=VersionCatalogOnly`has created both a libs.versions.toml and a versions.properties, with the same versions, no placeholders.

```
## Generated by $ ./gradlew refreshVersionsCatalog
[plugins]
org-jetbrains-kotlin-jvm = { id = "org.jetbrains.kotlin.jvm", version = "2.0.21" }
nu-studer-jooq = { id = "nu.studer.jooq", version = "9.0" }
org-flywaydb-flyway = { id = "org.flywaydb.flyway", version = "9.22.3" }
com-github-ben-manes-versions = { id = "com.github.ben-manes.versions", version = "0.51.0" }
[versions]
kotlin = "2.0.21"
kotlinx-coroutines = "1.9.0"
http4k = "5.32.4.0"
kotlinx-html = "0.11.0"
junit-jupiter = "5.11.3"
strikt = "0.35.1"
[libraries]
kotlinx-coroutines-core = { group = "org.jetbrains.kotlinx", name = "kotlinx-coroutines-core", version.ref = "kotlinx-coroutines" }
result4k = "dev.forkhandles:result4k:2.20.0.0"
slf4j-jdk14 = "org.slf4j:slf4j-jdk14:2.0.16"
http4k-bom = { group = "org.http4k", name = "http4k-bom", version.ref = "http4k" }
http4k-core = { module = "org.http4k:http4k-core" }
http4k-server-undertow = { module = "org.http4k:http4k-server-undertow" }
http4k-cloudnative = { module = "org.http4k:http4k-cloudnative" }
http4k-client-apache = { module = "org.http4k:http4k-client-apache" }
jackson-databind = "com.fasterxml.jackson.core:jackson-databind:2.18.0"
jackson-module-kotlin = "com.fasterxml.jackson.module:jackson-module-kotlin:2.18.0"
jackson-module-parameter-names = "com.fasterxml.jackson.module:jackson-module-parameter-names:2.18.0"
jackson-datatype-jdk8 = "com.fasterxml.jackson.datatype:jackson-datatype-jdk8:2.18.0"
jackson-datatype-jsr310 = "com.fasterxml.jackson.datatype:jackson-datatype-jsr310:2.18.0"
postgresql = "org.postgresql:postgresql:42.7.4"
hikaricp = "com.zaxxer:HikariCP:6.0.0"
jooq = "org.jooq:jooq:3.19.14"
kotlinx-html-jvm = { group = "org.jetbrains.kotlinx", name = "kotlinx-html-jvm", version.ref = "kotlinx-html" }
jooq-codegen = "org.jooq:jooq-codegen:3.19.1"
kotlin-build-tools-impl = { module = "org.jetbrains.kotlin:kotlin-build-tools-impl" }
kotlin-scripting-compiler-embeddable = { group = "org.jetbrains.kotlin", name = "kotlin-scripting-compiler-embeddable", version.ref = "kotlin" }
junit-jupiter-api = { group = "org.junit.jupiter", name = "junit-jupiter-api", version.ref = "junit-jupiter" }
junit-jupiter-params = { group = "org.junit.jupiter", name = "junit-jupiter-params", version.ref = "junit-jupiter" }
junit-jupiter-engine = { group = "org.junit.jupiter", name = "junit-jupiter-engine", version.ref = "junit-jupiter" }
kotlin-test = { module = "org.jetbrains.kotlin:kotlin-test" }
strikt-core = { group = "io.strikt", name = "strikt-core", version.ref = "strikt" }
playwright = "com.microsoft.playwright:playwright:1.48.0"
http4k-testing-approval = { module = "org.http4k:http4k-testing-approval" }
http4k-testing-hamkrest = { module = "org.http4k:http4k-testing-hamkrest" }
http4k-testing-strikt = { module = "org.http4k:http4k-testing-strikt" }
```

If I delete the version.properties it doesn't build, but as far as I can see the versions are now hard-coded in the toml?

**Reply by LouisCAD** [reply]
You can remove all the keys in `versions.properties`, but we haven't made refreshVersions work without it entirely yet.
